kernel/aster-nix/src/syscall/arch/x86.rs

```diff
@@ -0,0 +1,245 @@
+    execve::{sys_execve, sys_execveat},
+    exit::sys_exit,
+    exit_group::sys_exit_group,
+    fcntl::sys_fcntl,
+    fork::sys_fork,
+    fsync::sys_fsync,
+    futex::sys_futex,
+    getcwd::sys_getcwd,
+    getdents64::sys_getdents64,
+    getegid::sys_getegid,
+    geteuid::sys_geteuid,
+    getgid::sys_getgid,
+    getgroups::sys_getgroups,
+    getpeername::sys_getpeername,
+    getpgrp::sys_getpgrp,
+    getpid::sys_getpid,
+    getppid::sys_getppid,
+    getrandom::sys_getrandom,
+    getresgid::sys_getresgid,
+    getresuid::sys_getresuid,
+    getsid::sys_getsid,
+    getsockname::sys_getsockname,
+    getsockopt::sys_getsockopt,
+    gettid::sys_gettid,
+    gettimeofday::sys_gettimeofday,
+    getuid::sys_getuid,
+    ioctl::sys_ioctl,
+    kill::sys_kill,
+    link::{sys_link, sys_linkat},
+    listen::sys_listen,
+    lseek::sys_lseek,
+    madvise::sys_madvise,
+    mkdir::{sys_mkdir, sys_mkdirat},
+    mmap::sys_mmap,
+    mprotect::sys_mprotect,
+    munmap::sys_munmap,
+    open::{sys_open, sys_openat},
+    pause::sys_pause,
+    pipe::{sys_pipe, sys_pipe2},
+    poll::sys_poll,
+    prctl::sys_prctl,
+    pread64::sys_pread64,
+    prlimit64::sys_prlimit64,
+    read::sys_read,
+    readlink::{sys_readlink, sys_readlinkat},
+    recvfrom::sys_recvfrom,
+    rename::{sys_rename, sys_renameat},
+    rmdir::sys_rmdir,
+    rt_sigaction::sys_rt_sigaction,
+    rt_sigprocmask::sys_rt_sigprocmask,
+    rt_sigreturn::sys_rt_sigreturn,
+    sched_yield::sys_sched_yield,
+    select::sys_select,
+    sendto::sys_sendto,
+    set_get_priority::{sys_get_priority, sys_set_priority},
+    set_robust_list::sys_set_robust_list,
+    set_tid_address::sys_set_tid_address,
+    setfsgid::sys_setfsgid,
+    setfsuid::sys_setfsuid,
+    setgid::sys_setgid,
+    setgroups::sys_setgroups,
+    setpgid::sys_setpgid,
+    setregid::sys_setregid,
+    setresgid::sys_setresgid,
+    setresuid::sys_setresuid,
+    setreuid::sys_setreuid,
+    setsid::sys_setsid,
+    setsockopt::sys_setsockopt,
+    setuid::sys_setuid,
+    shutdown::sys_shutdown,
+    sigaltstack::sys_sigaltstack,
+    socket::sys_socket,
+    socketpair::sys_socketpair,
+    stat::{sys_fstat, sys_fstatat, sys_lstat, sys_stat},
+    statfs::{sys_fstatfs, sys_statfs},
+    symlink::{sys_symlink, sys_symlinkat},
+    sync::sys_sync,
+    syscall_dispatch, syscall_handler,
+    tgkill::sys_tgkill,
+    time::sys_time,
+    truncate::{sys_ftruncate, sys_truncate},
+    umask::sys_umask,
+    uname::sys_uname,
+    unlink::{sys_unlink, sys_unlinkat},
+    utimens::sys_utimensat,
+    wait4::sys_wait4,
+    waitid::sys_waitid,
+    write::sys_write,
+    writev::sys_writev,
+};
+
+syscall_dispatch! {
+    args, context, syscall_dispatch,
+
+    SYS_READ = 0 => syscall_handler!(3, sys_read, args),
+    SYS_WRITE = 1 => syscall_handler!(3, sys_write, args),
+    SYS_OPEN = 2 => syscall_handler!(3, sys_open, args),
```

Comment:
I think that the current design of this macro is acceptable although not ideal since it duplicates `syscall_handler` and `args` for every system call. I know Rust macros have their limitations and they may affect your ability to write macros with clear interfaces. But I am still wondering if it is possible to simplify the interface of macro `syscall_dispatch` to the following form or something similar.

```rust
// Implement the system call numbers and the system call dispatch function
// given the system numbers (the first column),
// the number of system call arguments (the second one),
// and the system call handler function (the third one).
impl_syscall_nums_and_dispatch_fn! {
    SYS_READ = 0      | 3 | sys_read ;
    SYS_WRITE = 1     | 3 | sys_write ;
    SYS_OPEN = 2      | 3 | sys_open ;
}
```

Reply:
I don't like putting the number of system call arguments separated, it's not intuitive though documented.

What about this one? It gives an intuition on how the arguments are passed.
``` rust
impl_syscall_nums_and_dispatch_fn! {
    SYS_READ = 0           => sys_read(args[..3]);
    SYS_WRITE = 1          => sys_write(args[..3]);
    SYS_OPEN = 2           => sys_open(args[..3]);
    SYS_RT_SIGRETURN = 15  => sys_rt_sigreturn(args[..0], context);
}
```

Reply:
I am ok with the new interface.